pkg/operator/sync_openshiftcontrollermanager_v311_00.go

```diff
@@ -281,10 +323,52 @@ func manageRouteControllerManagerClientCA_v311_00_to_latest(client coreclientv1.
 }
 
+// groupDisabledControllers gets caps from configv1.ClusterVersion and populates
+// groupDisabled with all controllers that are known in the ClusterVersion and
+// also disabled.
+func groupDisabledControllers(clusterVersion *configv1.ClusterVersion) map[controlplanev1.OpenShiftControllerName]bool {
+	knownCaps := sets.New[configv1.ClusterVersionCapability](clusterVersion.Status.Capabilities.KnownCapabilities...)
+	capsEnabled := sets.New[configv1.ClusterVersionCapability](clusterVersion.Status.Capabilities.EnabledCapabilities...)
+
+	groupDisabled := make(map[controlplanev1.OpenShiftControllerName]bool)
+	for _, cont := range knownControllers {
+		if controllerCapabilities[cont] != "" && knownCaps.Has(controllerCapabilities[cont]) && !capsEnabled.Has(controllerCapabilities[cont]) {
+			groupDisabled[cont] = true
+		}
+	}
+	return groupDisabled
+}
+
+// By default, all controllers are enabled. To disable specific controllers,
+// '-foo' disables the controller named 'foo'. For example, to disable the
+// OpenShiftServiceAccountController, one would use "-openshift.io/serviceaccount".
+func disableControllers(groupDisabled map[controlplanev1.OpenShiftControllerName]bool) []string {
+	controllers := []string{"*"}
+	for _, cont := range knownControllers {
+		if groupDisabled[cont] {
+			controllers = append(controllers, fmt.Sprintf("-%s", cont))
+		}
+	}
+	return controllers
+}
+
 // similar logic for route-controller-manager in manageRouteControllerManagerConfigMap_v311_00_to_latest
-func manageOpenShiftControllerManagerConfigMap_v311_00_to_latest(kubeClient kubernetes.Interface, client coreclientv1.ConfigMapsGetter, recorder events.Recorder, operatorConfig *operatorapiv1.OpenShiftControllerManager) (*corev1.ConfigMap, bool, error) {
+func manageOpenShiftControllerManagerConfigMap_v311_00_to_latest(clusterVersionInformer configinformersconfigv1.ClusterVersionInformer, kubeClient kubernetes.Interface, client coreclientv1.ConfigMapsGetter, recorder events.Recorder, operatorConfig *operatorapiv1.OpenShiftControllerManager) (*corev1.ConfigMap, bool, error) {
+	clusterVersion, err := clusterVersionInformer.Lister().Get("version")
+	if err != nil {
+		return nil, false, err
+	}
+	groupDisabled := groupDisabledControllers(clusterVersion)
 	configMap := resourceread.ReadConfigMapV1OrDie(v311_00_assets.MustAsset("v3.11.0/openshift-controller-manager/cm.yaml"))
 	ocmDefaultConfig := v311_00_assets.MustAsset("v3.11.0/config/openshift-controller-manager-defaultconfig.yaml")
+	openShiftControllerManagerConfig := &controlplanev1.OpenShiftControllerManagerConfig{}
+	yaml.Unmarshal(ocmDefaultConfig, openShiftControllerManagerConfig)
+	controllers := disableControllers(groupDisabled)
+	openShiftControllerManagerConfig.Controllers = controllers
+	ocmDefaultConfig, err = yaml.Marshal(openShiftControllerManagerConfig)
```

Comment:
When I do Unmarshal and Marshal with the new controllers slice, I obviously get a bunch of fields that were not set in the yaml before, all zero'ed out or nilled out. Like

```
{"apiVersion":"openshiftcontrolplane.config.openshift.io/v1","build":{"additionalTrustedCA":"","buildDefaults":{"resources":{}},"buildOverrides":null,"imageTemplateFormat":{"format":"quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:0007c01aaa6e2c68d1d82e2bd7d73e5d6d3fc457e11b569ea6c369a1532eb2a0","latest":false}},"controllers":["*","-openshift.io/build","-openshift.io/deploymentconfig"],"deployer":{"imageTemplateFormat":{"format":"quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:068af9903dc5419d5eafe8de1f93e4375566315dbad5df4f4d118bc46fdc63e2","latest":false}},"dockerPullSecret":{"internalRegistryHostname":"image-registry.openshift-image-registry.svc:5000","registryURLs":null},"featureGates":["BuildCSIVolumes=false"],"imageImport":{"disableScheduledImport":false,"maxScheduledImageImportsPerMinute":0,"scheduledImageImportMinimumIntervalSeconds":0},"ingress":{"ingressIPNetworkCIDR":""},"kind":"OpenShiftControllerManagerConfig","kubeClientConfig":{"connectionOverrides":{"acceptContentTypes":"","burst":0,"contentType":"","qps":0},"kubeConfig":""},"leaderElection":{"leaseDuration":"0s","renewDeadline":"0s","retryPeriod":"0s"},"network":{"clusterNetworks":null,"networkPluginName":"","serviceNetworkCIDR":"","vxlanPort":0},"resourceQuota":{"concurrentSyncs":0,"minResyncPeriod":"0s","syncPeriod":"0s"},"securityAllocator":{"mcsAllocatorRange":"","mcsLabelsPerProject":0,"uidAllocatorRange":""},"serviceAccount":{"managedNames":null},"serviceServingCert":{"signer":null},"servingInfo":null}
```

OCM sets most important defaults here: https://github.com/openshift/openshift-controller-manager/blob/80c4923e0b2014ec8b8a6845c21078e6cc036f91/pkg/cmd/openshift-controller-manager/openshiftcontrolplane_default.go#L12#L56

But I am just wondering if this is all ok

Reply:
You don't need to Unmarshal and Marshal back if you do:
```go
openShiftControllerManagerConfig := &controlplanev1.OpenShiftControllerManagerConfig{}
controllers := disableControllers(groupDisabled)
openShiftControllerManagerConfig.Controllers = controllers
bytes, err := yaml.Marshal(openShiftControllerManagerConfig)
...
requiredConfigMap, _, err := resourcemerge.MergeConfigMap(configMap, "config.yaml", nil, ocmDefaultConfig, bytes, operatorConfig.Spec.ObservedConfig.Raw, operatorConfig.Spec.UnsupportedConfigOverrides.Raw)
...
```
`resourcemerge.MergeConfigMap` will do the merging for you.

Reply:
we went this this: https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/291/commits/4eaa994bf58b1cc4804ddd0986ac484dcb131333